File: desktop/engine/paster.py
"""Cross-platform text paste into active application."""
import platform
import subprocess
import time

PLATFORM = platform.system()


def paste_text(text):
    """Copy text to clipboard and simulate paste keystroke."""
    if not text:
        return

    if PLATFORM == "Darwin":
        _paste_mac(text)
    elif PLATFORM == "Windows":
        _paste_windows(text)
    else:
        _paste_linux(text)


def replace_text(old_length, new_text):
    """Select and replace previously pasted text.

    Used for final correction: select back `old_length` chars, then paste new text.
    """
    if not new_text:
        return

    if PLATFORM == "Darwin":
        # Select back old_length characters via Shift+Left, then paste
        script = f'''
        tell application "System Events"
            repeat {old_length} times
                key code 123 using shift down
            end repeat
        end tell
        '''
        subprocess.run(["osascript", "-e", script], check=False)
        time.sleep(0.05)
        _paste_mac(new_text)

    elif PLATFORM == "Windows":
        from pynput.keyboard import Controller, Key
        kb = Controller()
        for _ in range(old_length):
            kb.press(Key.left)
            kb.release(Key.left)
        time.sleep(0.05)
        _paste_windows(new_text)

    else:
        _paste_linux(new_text)


def _paste_mac(text):
    """macOS: pbcopy + Cmd+V via osascript."""
    proc = subprocess.Popen(["pbcopy"], stdin=subprocess.PIPE)
    proc.communicate(text.encode("utf-8"))
    subprocess.run([
        "osascript", "-e",
        'tell application "System Events" to keystroke "v" using command down'
    ], check=False)


def _paste_windows(text):
    """Windows: clip.exe + Ctrl+V via pynput."""
    proc = subprocess.Popen(["clip.exe"], stdin=subprocess.PIPE)
    proc.communicate(text.encode("utf-16-le"))
    from pynput.keyboard import Controller, Key
    kb = Controller()
    kb.press(Key.ctrl)
    kb.press('v')
    kb.release('v')
    kb.release(Key.ctrl)


def _paste_linux(text):
    """Linux: xclip + Ctrl+V via pynput."""
    proc = subprocess.Popen(["xclip", "-selection", "clipboard"], stdin=subprocess.PIPE)
    proc.communicate(text.encode("utf-8"))
    from pynput.keyboard import Controller, Key
    kb = Controller()
    kb.press(Key.ctrl)
    kb.press('v')
    kb.release('v')
    kb.release(Key.ctrl)

File: rs/src/paster.rs
/// Cross-platform text paste into active application.
///
/// Port of python/engine/paster.py.
/// Mac: Cmd+V, Windows/Linux: Ctrl+V.

use arboard::Clipboard;
use std::thread;
use std::time::Duration;

/// Copy text to clipboard and simulate paste keystroke.
pub fn paste_text(text: &str) {
    if text.is_empty() {
        return;
    }

    // Copy to clipboard
    match Clipboard::new() {
        Ok(mut clipboard) => {
            if let Err(e) = clipboard.set_text(text.to_string()) {
                eprintln!("[typeoff] Failed to set clipboard: {}", e);
                return;
            }
        }
        Err(e) => {
            eprintln!("[typeoff] Failed to open clipboard: {}", e);
            return;
        }
    }

    thread::sleep(Duration::from_millis(50));

    // Simulate paste keystroke
    // Each call creates a fresh Enigo — avoids stale state issues
    #[cfg(target_os = "macos")]
    paste_macos();

    #[cfg(target_os = "windows")]
    paste_windows();

    #[cfg(target_os = "linux")]
    paste_linux();
}

#[cfg(target_os = "macos")]
fn paste_macos() {
    // Small delay to let the target app regain focus after any window changes
    thread::sleep(Duration::from_millis(100));

    // Use osascript to send Cmd+V to the frontmost app
    // This requires Accessibility permission for System Events
    let _ = std::process::Command::new("osascript")
        .arg("-e")
        .arg("tell application \"System Events\" to keystroke \"v\" using command down")
        .output();
}

#[cfg(target_os = "windows")]
fn paste_windows() {
    use enigo::{Enigo, Key, Keyboard, Settings};
    if let Ok(mut enigo) = Enigo::new(&Settings::default()) {
        let _ = enigo.key(Key::Control, enigo::Direction::Press);
        let _ = enigo.key(Key::Unicode('v'), enigo::Direction::Click);
        let _ = enigo.key(Key::Control, enigo::Direction::Release);
    }
}

#[cfg(target_os = "linux")]
fn paste_linux() {
    use enigo::{Enigo, Key, Keyboard, Settings};
    if let Ok(mut enigo) = Enigo::new(&Settings::default()) {
        let _ = enigo.key(Key::Control, enigo::Direction::Press);
        let _ = enigo.key(Key::Unicode('v'), enigo::Direction::Click);
        let _ = enigo.key(Key::Control, enigo::Direction::Release);
    }
}
